js/tests/unit/scrollspy.js

```diff
@@ -142,6 +142,48 @@ $(function () {
       .then(function () { return testElementIsActiveAfterScroll('#li-2', '#div-2') })
   })
 
+  QUnit.test('should add the active class correctly when there are nested elements at 0 scroll offset', function (assert) {
+    var navbarHtml =
+        '<nav class="navbar">'
+      + '<ul class="nav">'
+      + '<li id="li-1"><a href="#div-1">div 1</a>'
+      + '<ul>'
+      + '<li id="li-2"><a href="#div-2">div 2</a></li>'
+      + '</ul>'
+      + '</li>'
+      + '</ul>'
+      + '</nav>'
+    var contentHtml =
+        '<div class="content" style="position: absolute; top: 0px; overflow: auto; height: 50px">'
+      + '<div id="div-1" style="padding: 0; margin: 0">'
+      + '<div id="div-2" style="height: 200px; padding: 0; margin: 0">div 2</div>'
+      + '</div>'
+      + '</div>'
+
+    $(navbarHtml).appendTo('#qunit-fixture')
+    var $content = $(contentHtml)
+      .appendTo('#qunit-fixture')
+      .bootstrapScrollspy({ offset: 0, target: '.navbar' })
+
+    var testElementsAreActiveAfterScroll = function () {
+      var deferred = $.Deferred()
+      var scrollHeight = $content.scrollTop() + 10
+      var done = assert.async()
+      $content.one('scroll', function () {
+        assert.ok($('#li-1').hasClass('active'), 'nav item for outer element has "active" class')
+        assert.ok($('#li-2').hasClass('active'), 'nav item for inner element has "active" class')
+        done()
+        deferred.resolve()
+      })
+      $content.scrollTop(scrollHeight)
+      return deferred.promise()
+    }
+
+    $.when(testElementsAreActiveAfterScroll())
+      .then(testElementsAreActiveAfterScroll)
+      .then(testElementsAreActiveAfterScroll)
```

Comment:
Hi! I'm having a little confusion over this. 

The first confusion is that the test in your gist only runs the assertions once. Since on the first pass `++times == 1` and the second pass `++times == 2`, which triggers `done()` before doing an iteration.

When I change the condition to `== 4` then the test iterates three times, as the one in this commit does. However, all the assertions pass as expected — I can't seem to reproduce the failure you mention.

I've made a new branch with both tests in, and some hacky logging to show the value of the assertion and the `scrollTop`. Both tests seem to generate the same output. [Here's the branch.](https://github.com/neoeno/bootstrap/tree/handle_multiple_zero_offset_scrollspy_elements_2)

Am I missing something?

I actually think your test, using recursion, looks a fair bit cleaner than this approach using promises. Perhaps at the expense of some readability (for me), but I'm definitely open to either option. I just copied and rewrote the "should add the active class correctly when there are nested elements at 0 scroll offset".